src/java/com/zimbra/cs/taglib/tag/SendTFACodeTag.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
+ * Zimbra Collaboration Suite Server
+ * Copyright (C) 2023 Zimbra, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software Foundation,
+ * version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License along with this program.
+ * If not, see <http://www.gnu.org/licenses/>.
+ * ***** END LICENSE BLOCK *****
+ */
+package com.zimbra.cs.taglib.tag;
+
+import java.io.IOException;
+
+import javax.servlet.jsp.JspContext;
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.PageContext;
+
+import com.zimbra.client.ZAuthResult;
+import com.zimbra.common.auth.ZAuthToken;
+import com.zimbra.common.service.ServiceException;
+import com.zimbra.common.soap.AccountConstants;
+import com.zimbra.common.soap.Element;
+import com.zimbra.common.soap.Element.XMLElement;
+import com.zimbra.common.soap.SoapHttpTransport;
+import com.zimbra.common.util.ZimbraLog;
+import com.zimbra.cs.taglib.ZJspSession;
+import com.zimbra.cs.taglib.bean.BeanUtils;
+import com.zimbra.soap.account.message.SendTwoFactorAuthCodeRequest.SendTwoFactorAuthCodeAction;
+import com.zimbra.soap.account.message.SendTwoFactorAuthCodeResponse.SendTwoFactorAuthCodeStatus;
+
+public class SendTFACodeTag extends ZimbraSimpleTag {
+
+    private String mVarResult;
+    private String mMethod;
+    private ZAuthResult authResult;
+
+    public void setVarResult(String varResult) { this.mVarResult = varResult; }
+    public void setMethod(String method) { this.mMethod = method; }
+    public void setAuthResult(ZAuthResult authResult) { this.authResult = authResult; }
+
+    public void doTag() throws JspException, IOException {
+        String action = "";
+        try {
+            JspContext jctxt = getJspContext();
+            PageContext pageContext = (PageContext) jctxt;
+            ZAuthToken zAuthToken;
+            String authtoken;
+
+            if (authResult != null) {
+                // no authtoken is included in a http request just after login with username and password.
+                // authtoken needs to be fetched from ZAuthResult
+                zAuthToken = authResult.getAuthToken();
+            } else {
+                // ZAuthResult is empty when a TFA method is changed on login page.
+                // check authtoken in a http request
+                zAuthToken = ZJspSession.getAuthToken(pageContext);
+            }
+
+            if (!BeanUtils.isTFARequired(zAuthToken)) {
+                ZimbraLog.webclient.warn("Invalid authtoken was used to call SendTFACodeRequest");
```

Comment:
For consistency, we should refactor SendTFACodeRequest to SendTwoFactorAuthCodeRequest. Also we can refactor class name if required. `SendTFACodeTag` and `GetTFAConfigTag`